### PR TITLE
feat(window/ui): add splitkeep option 

### PR DESCRIPTION
**Pull request by luukvbaal**
Hi, I'm the author of a relatively popular [plugin](https://github.com/luukvbaal/stabilize.nvim). I'm looking for feedback on whether something like this PR has any possibility of being merged with the behavior of said plugin being provided in core behind a new option.

I had originally assumed while writing the plugin that the "jarring" phenomenon of window viewports changing(scrolling), when splitting windows horizontally, was some undesired side effect of managing the windows. Now, having finally taken the time to look into the source code, it seems this is done intentionally with the following justification:
https://github.com/neovim/neovim/blob/41785b1b0cd8b54700110d0c83f599126b38a8c9/src/nvim/window.c#L6193-L6194

After 2 months of work on the vim patch it was finally merged. This PR now ports https://github.com/vim/vim/commit/29ab524358ba429bcf6811710afc97a978641f0b, https://github.com/vim/vim/commit/5ed391708a62b4ebaa84dd23e32a416e5c3383d9, https://github.com/vim/vim/commit/470a14140bc06f1653edf26ab0b3c9b801080353, https://github.com/vim/vim/commit/3735f11050616652525bf80b4fbcb2b3bfeab113, https://github.com/vim/vim/commit/594f9e09cd68e6277b8aa08094405bc642c5792a, https://github.com/vim/vim/commit/d5bc762dea1fd32fa04342f8149f95ccfc3b9709, https://github.com/vim/vim/commit/faf1d412f5e3665021500b528c0e7301eb02bf0b, https://github.com/vim/vim/commit/20e58561abc4116f3bfbafaef242d886dd77b303, https://github.com/vim/vim/commit/7c1cbb6cd437c6e0c3ccc05840cc931108b4a60a, https://github.com/vim/vim/commit/439a2ba1749463718b6ce1e1375b68c7b7cff808, https://github.com/vim/vim/commit/13ece2ae1d09009d3fb8acf858c288e7848ecdac, https://github.com/vim/vim/commit/346823d3e5668b99d2c2fd920e7f215e21ad3ea7.

### CODE REVIEW

**Comment by luukvbaal** [comment]
BTW the motivation behind trying to merge it into core is that the "stabilization" done by the plugin is sometimes not fast enough in faster terminals. Resulting in flickering when restoring the topline.

**Comment by lervag** [comment]
I'm not a neovim dev, but I have some immediate thoughts:

* First, this is cool, I would be happy to see it merged!
* I think you should also add documentation. E.g. in `runtime/doc/quickref.txt` and `runtime/doc/options.txt`.
* Would it make sense to add a test? Not sure how to write such a test, though.

**Comment by luukvbaal** [comment]
>     * I think you should also add documentation. E.g. in `runtime/doc/quickref.txt` and `runtime/doc/options.txt`.

Definitely. I refrained from doing so as I was still looking for general feedback on the feasibility of the PR. The PR turned out to be easier than I imagined though so I guess I can draft the help docs.

>     * Would it make sense to add a test? Not sure how to write such a test, though.

Not sure about this either.

I'm also wondering about the etiquette of opening PRs agains vim. I have no interest in doing so myself but I quickly checked and this part of core vim seems similar to nvim still.

**Comment by clason** [comment]
> I'm also wondering about the etiquette of opening PRs agains vim. I have no interest in doing so myself but I quickly checked and this part of core vim seems similar to nvim still.

The etiquette is very much to contribute to Vim, if at all appropriate :) 

In this case, you are right that this PR is directly applicable to Vim. The best course of action is then indeed to first make the PR to Vim, see what they think, and if they merge it, port that patch.

This way both projects profit from the new feature, and we avoid needless divergence in case Vim decides on slight changes to, say, the abbreviated option name.

(If Vim rejects the feature, we can reopen this PR, now with the knowledge that we don't need to take upstream into account.)

**Comment by luukvbaal** [comment]
> The etiquette is very much to contribute to Vim,

I guess I'll do that then.

**Comment by justinmk** [comment]
- This is definitely wanted and will likely become a Nvim default, so thank you.
- Also makes sense to see if upstream Vim wants this, please let us know the result of that.

Adding a screen test is easy, see [test/functional/example_spec.lua](https://github.com/neovim/neovim/blob/master/test/functional/example_spec.lua) for example. You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) like this:

    TEST_FILE=test/functional/example_spec.lua make functionaltest

I didn't find an existing test file where this would obviously fit, but maybe one of these:

```
test/functional/ui/screen.lua
test/functional/ui/screen_basic_spec.lua
test/functional/ui/multigrid_spec.lua
```

else we might need a new file, `test/functional/ui/window_spec.lua`.

**Comment by lewis6991** [comment]
Are you sure we would want this as default? Isn't the point of the scrolling to ensure the split doesn't cover the last position of the cursor?

**Comment by justinmk** [comment]
> Isn't the point of the scrolling to ensure the split doesn't cover the last position of the cursor?

The current `:split` behavior scrolls to keep the cursor at the same relative position, e.g. 30% or 50%. 

I would expect 'splitscroll=false' to scroll if the cursor position would be hidden, since there is not yet an internal concept of a window with an off-screen cursor.

(btw, does this imply we need a number or flags option instead of a boolean? because there are several possible behaviors.)

**Comment by lewis6991** [comment]
Hmm, sounds like the current behaviour is naive/simple and now we just want to make it "smart". I'd contemplate not even making an option for this and just change the current behaviour, unless there's a good argument I'm not realising.

> (btw, does this imply we need a number or flags option instead of a boolean? because there are several possible behaviors.)

Can you list these?

Also note we need to make sure we always respect `'scrolloff'`.

**Comment by qubidt** [comment]
> Hmm, sounds like the current behaviour is naive/simple and now we just want to make it "smart"

Is this PR not just allowing currently existing "smart" behavior to be disabled?

**Comment by luukvbaal** [comment]
> (btw, does this imply we need a number or flags option instead of a boolean? because there are several possible behaviors.)
 Yes this makes sense, the option was also available in the plugin. There it was named `force` as in "force stabilizing even when cursor is behind newly opened window". Not sure what an apt flag name would be if the option name stays `splitscroll`. Perhaps it should be reconsidered?

**Comment by justinmk** [comment]
> Not sure what an apt flag name would be if the option name stays `splitscroll`. Perhaps it should be reconsidered?

could just be a number option, see 'laststatus'.

> > (btw, does this imply we need a number or flags option instead of a boolean? because there are several possible behaviors.)
> 
> Can you list these?

1. scroll to maintain relative position (current behavior)
2. don't scroll except as needed to keep cursor row visible (and subject to `'scrolloff'`)
3. never scroll, even if cursor row will be hidden or 'scrolloff' would be violated

**Comment by luukvbaal** [comment]
> Are you sure we would want this as default? Isn't the point of the scrolling to ensure the split doesn't cover the last position of the cursor? I'd contemplate not even making an option for this and just change the current behaviour

Even when the cursor would not be hidden behind the split, currently scrolling happens just to "keep the same relative position". To me this is necessary but I guess it could be up to personal preference. Hence the option, but perhaps not necessary if we think the current default is undesired.

**Comment by justinmk** [comment]
Keeping the relative position is a valid use-case. Could avoid a new option by changing the core behavior, then users who want "keep relative position" could use a `WinScroll` (which apparently also means "WinResize"...?) event, but idk if `WinScroll`/`WinResize` gives the necessary info in `v:event` yet. 

But that doesn't resolve the question about letting the cursor row get hidden: that does not seem like a good default behavior, and raises a lot of questions.

**Comment by luukvbaal** [comment]
I thought it could look something like the latest commit. Perhaps the solution is too naive though, I never use `scrolloff` so I'm not sure what to consider.

**Comment by luukvbaal** [comment]
Vim upstream will consider the feature if we can agree on what it will do. They will require a test as well, `testdir/test_window_cmd.vim` seems appropriate. Will adding a test there be sufficient or are the legacy tests deprecated here?

**Comment by zeertzjq** [comment]
> Will adding a test there be sufficient

Yes

**Comment by lewis6991** [comment]
> Keeping the relative position is a valid use-case. Could avoid a new option by changing the core behavior, then users who want "keep relative position" could use a WinScroll (which apparently also means "WinResize"...?) event, but idk if WinScroll/WinResize gives the necessary info in v:event yet.

Ok, let's just introduce an option.

> But that doesn't resolve the question about letting the cursor row get hidden: that does not seem like a good default behavior, and raises a lot of questions. If we are going to entertain the idea of letting the cursor row get hidden, that would need to be a third choice.

I'm working on the assumption that any new behaviour we add, **doesn't** break `'scrolloff'`, and, **never** hides (or moves) the cursor. For that reason I don't think we should add a 3rd form of behaviour.

In terms of how we do it, I think we should just add `'splitoptions'`, even if it only contains one option.


> I thought it could look something like the latest commit. Perhaps the solution is too naive though, I never use scrolloff so I'm not sure what to consider.

We need to check if the cursor will remain in the window and obeys `'scrolloff'`. If it doesn't then we need to scroll appropriately, and not necessarily the old scroll behaviour, but instead the minimum amount.

**Comment by luukvbaal** [comment]
> For that reason I don't think we should add a 3rd form of behaviour.

I don't see any harm in adding the third option for those who prefer it. It was the default for my plugin. I agree that the cursor position should never violate `scrolloff` for the active window However, I think delaying the scrolling to when the cursor is moved back the window where scrolling would be required to honor `scrolloff` (current behavior of this PR for `splitscroll=2`), results in better UX.

> but instead the minimum amount.

Yeah I considered this and I agree it would probably be better behavior but I wasn't sure it would be worth the hassle.

**Comment by luukvbaal** [comment]
> We need to check if the cursor will remain in the window and obeys `'scrolloff'`. If it doesn't then we need to scroll appropriately, and not necessarily the old scroll behaviour, but instead the minimum amount.

Latest commit does this when `'splitscroll' == 1` by setting `w_fraction` proportional to  `'scrolloff'` before the `scroll_to_fraction()` call. This results in the minimum amount scrolled required to obey `'scrolloff'`. 

Caveat being that this condition should actually be `w_botline` rather than `w_topline + height`. The problem is that `w_botline` is not yet updated at this point after opening a split . I'm not sure what is required to update `w_botline` at this point (and if what is required will be worth it for this edge case). Hence the condition is not accurately evaluated when wrapped or folded lines are present, possibly resulting in the cursor being hidden behind the split:
https://github.com/neovim/neovim/blob/75a073bf04455da6d7e310b2e68cea454518a2f8/src/nvim/window.c#L6326-L6327

**Comment by luukvbaal** [comment]
> In terms of how we do it, I think we should just add 'splitoptions', even if it only contains one option.

> > For that reason I don't think we should add a 3rd form of behaviour.

> I don't see any harm in adding the third option for those who prefer it

Also wondering if there's any other feedback on the general shape of the PR regarding the previous comments. Haven't heard any strong opionions from any Vim devs.

**Comment by luukvbaal** [comment]
> I'm not sure what is required to update w_botline at this point (and if what is required will be worth it for this edge case).

A `redraw_asap()` seems to work but I don't know if anything less expensive might be sufficient as well.

**Comment by justinmk** [comment]
> In terms of how we do it, I think we should just add `'splitoptions'`, even if it only contains one option.

Looks like the 'xxoptions' convention is for a list of boolean flags. Whereas this option is a enum-style option, i.e. one of many mutually-exclusive choices.

lgtm, lmk if there is no movement on Vim side.

File: runtime/doc/options.txt
Comment (justinmk) [inline runtime/doc/options.txt:5924]: it doesn't "influence", it decides. Also don't need verbose preamble like "The value of this option...".

```suggestion
	Decides the scroll behavior when opening
```

Comment (luukvbaal) [reply runtime/doc/options.txt:5924]: I would tend to agree but I tried to mimic the wording of other options:
https://github.com/neovim/neovim/blob/b78e4139b2914ed0e35f7643461561683c8051b7/runtime/doc/options.txt#L3657-L3664

**Comment by luukvbaal** [comment]
Vim is requesting user experience with the patch: https://github.com/vim/vim/pull/10682#issuecomment-1193060554.

**Comment by luukvbaal** [comment]
Cross-posting https://github.com/vim/vim/pull/10682#issuecomment-1208727193 for feedback. 

> don't scroll except as needed to keep cursor row visible (and subject to 'scrolloff')

@justinmk @lewis6991 do we feel this case is necessary? I tackled the `neverscroll` case first while reworking the option and now I'm not sure it will be worth the hassle/code to bring it back. The whole point of the option is to keep the windows stable. The old cursor position will be available in either the old or new window, it's easy to find one's way back there.

**Comment by luukvbaal** [comment]
> The old cursor position will be available in either the old or new window, it's easy to find one's way back there.

I guess this is only true when no split arg is present. Perhaps the option still makes sense for that case. Also, I think the only case where we do want to scroll instead of setting the cursor position, regardless of the option, is when the cursor will be invalid and we are not in normal mode. Not handled yet in the latest vim PR commit but we don't want to update the cursor position while in insert mode for example, should scroll in that case.

**Comment by luukvbaal** [comment]
https://github.com/vim/vim/pull/10682#issuecomment-1211295521

> Alright, I'm happy with how this looks and works now. I would consider it mergeable and would put it up for review brammool but I haven't had any feedback yet from neovim side on the omission of the noscroll option(or here for that matter, but there were some opinions shared on the option earlier and it originated over at the neovim PR). The addition of the jumplist entry with the old cursor position makes it even more obsolete imo.
The failing test on appveyor with gvim still requires fixing I guess but I haven't been able to reproduce it yet. It seems to be running the test on windows, perhaps I can reproduce it there.

**Comment by luukvbaal** [comment]
Bram is again asking for feedback on the vim patch. It has changed quite a bit since last time. We now always avoid scrolling, for both windows split above and below, save the previous cursor position in the jumplist when we change the cursor position to avoid scrolling, and catch some edge cases concerning tabline and laststatus. Still wondering how people feel about the second option of the previously proposed flags:

>
    1. scroll to maintain relative position (current behavior)
    2. don't scroll except as needed to keep cursor row visible (and subject to 'scrolloff')
    3. never scroll, even if cursor row will be hidden or 'scrolloff' would be violated

Could try to add it back but I'm not sure we need it now. The whole point of the option is to avoid scrolling and we add the previous cursor position to the jumplist @justinmk @lewis6991. I guess some people might want to take advantage of the no scrolling behavior only in the second case though.

**Comment by lewis6991** [comment]
Makes sense to me.

**Comment by luukvbaal** [comment]
Rebased in the hope that anyone might want to test out a neovim build instead of the vim patch. Please provide feedback in https://github.com/vim/vim/pull/10682 if you have any.

**Comment by luukvbaal** [comment]
> Makes sense to me.

Sorry @lewis6991, but what makes sense to you? Omitting the second option, or trying to add it back? I was scrolling through the previous feedback and you emphasized being strongly against ever altering the cursor position: https://github.com/neovim/neovim/pull/19243#issuecomment-1180114717

>I'm working on the assumption that any new behaviour we add, doesn't break 'scrolloff', and, never hides (or moves) the cursor.

The current PR violates that, did your opinion change on the matter or did you mean adding the option back would make sense. I don't think it's that big of a deal. We save the previous cursor position and make sure to not violate 'scrolloff'. In exchange, we can make the option work, as in stabilize window content, for both 'above' and 'below' splits, regardless of the cursor postion. In any case, I think we need more feedback on whether the `noscroll` option is deemed necessary. I wouldn't use it and keeping the option simple also has it's benefits so looking for other opinions.

**Comment by lewis6991** [comment]
Makes sense to omit the second option.

You've clearly put a lot more thought into this than I have, and it seems vim is also happy to not scroll. And saving the cursor position seems like a good compromise.

**Comment by luukvbaal** [comment]
Worth noting that some edge case handling in the vim patch regarding winbar/tabline don't seem to be necessary here due to Neovims inner window size. On the other hand, for Neovim we require some checks for floating windows that don't seem to be necessary for vims popup windows. I'll update accordingly once the vim patch is finalized and merged (or rejected).

File: src/nvim/window.c
Comment (justinmk) [inline src/nvim/window.c:6175]: - `fix_cursor` is an existing, less verbose precedent.
- "spsc" is pretty obscure. "win_" is a common prefix that fits here.

```suggestion
static void win_fix_scroll(win_T *next_curwin, int flags)
```

File: src/nvim/window.c
Comment (justinmk) [inline src/nvim/window.c:6247]: The docstring should explain how this is different from `fix_cursor`.

```suggestion
static void win_fix_cursor(win_T *wp, linenr_T lnum, int curnormal)
```

File: src/nvim/window.c
Comment (justinmk) [inline src/nvim/window.c:6274]: missing blank line

File: src/nvim/window.c
Comment (justinmk) [inline src/nvim/window.c:6270]: this could be 0. is that expected?

Comment (luukvbaal) [reply src/nvim/window.c:6270]: Thanks for reviewing, I'll adopt the changes in the vim patch for now.
What makes you say it could be 0, I'm guarding against that above with `MAX((linenr_T)(wp->w_botline - so - 1), 1);` or am I missing something?

Comment (justinmk) [reply src/nvim/window.c:6270]: I missed the `if (nlnum)` check, sorry!

**Comment by luukvbaal** [comment]
Bram has included the patch, I'll work on porting it.

**Comment by luukvbaal** [comment]
Latest commit ports, https://github.com/vim/vim/commit/29ab524358ba429bcf6811710afc97a978641f0b, https://github.com/vim/vim/commit/5ed391708a62b4ebaa84dd23e32a416e5c3383d9 and https://github.com/vim/vim/pull/11122. Some problems are still popping up now that the feature is tested more thoroughly by vim users. So perhaps its worthwhile to wait a bit longer, but I thought I would update the PR for feedback/testing here.

It took a great effort (for me) to get the feature right so keep in mind that the vim patch has seen a lot more testing by me over the past weeks. I don't see any immediate problems though with this basically one-to-one port. Only changes are skipping floating windows while looping over all windows in win_fix_scroll() (not necessary for vim's popup windows as they are stored in a separate list. As well as not accounting for winbar in win_init() while copying over botline, height and winrow. Other than that the location of some win_fix_scroll() calls is different due to neovim's inner window changes to window.c. I'll go over the changes myself one more time and see if some older feedback here still applies. In the meantime, the PR is updated.

**Comment by luukvbaal** [comment]
I see now I have to familiarize myself with the neovim test suite. I ran the test locally and it passed but there were similar problems that only showed up in vim's workflows. I'll look into it.

**Comment by luukvbaal** [comment]
In porting https://github.com/vim/vim/commit/470a14140bc06f1653edf26ab0b3c9b801080353 there appears to be an incompatibility with vim. The 'scroll' test fails when a winbar is present, i.e. 'scroll' is not half the window height. To fix this we can either account for this in the test, or make neovim compatible again by using `w_height_inner` in, `win_comp_scroll()`:
```diff
@@ -6367,7 +6367,7 @@ void win_comp_scroll(win_T *wp)
 {
   const long old_w_p_scr = wp->w_p_scr;

-  wp->w_p_scr = wp->w_height / 2;
+  wp->w_p_scr = wp->w_height_inner / 2;
   if (wp->w_p_scr == 0) {
     wp->w_p_scr = 1;
   }
```
The latter seems more appropriate to me as I think it would make sense for 'scroll' to correlate to the inner window height.

File: src/nvim/window.c
Comment (zeertzjq) [reply src/nvim/window.c:6369]: The options code look fine. `if_false` only has effect when `condition` is present.

Comment (luukvbaal) [reply src/nvim/window.c:6369]: > So this fixes the splitscroll test

It appears it still fails somewhere on CI, couldn't reproduce it locally anymore.

Comment (luukvbaal) [reply src/nvim/window.c:6370]: > I suspect it has something to do with default option values.

Right, we have to unset `winbar` as it is a global option for nvim. As opposed to vim's window local winbar. Fixed in latest commit.

(I'm aware we can `setlocal winbar` but this would result in the winbar disappearing as soon as the window is closed, which happens a lot in this test. As such, we are not really testing the same window layouts as in vim, as each window will have a winbar. But at least we are confirming that no scrolling happens with winbar enabled.)

**Comment by luukvbaal** [comment]
Right so latest commit basically completes the port, including as of yet unmerged https://github.com/vim/vim/pull/11134. Don't know if we want to wait some more to see if any more problems show up with vim users testing the option, or merge here(after review of course) to get even more feedback, but the few kinks that showed up in porting the patch are resolved as far as I can tell.

There is still https://github.com/vim/vim/commit/5ed391708a62b4ebaa84dd23e32a416e5c3383d9#commitcomment-84011477, and I see it takes even longer on neovim CI: 120 seconds for some [platforms](https://github.com/neovim/neovim/actions/runs/3056880192/jobs/4931459242#step:14:4142), whoops!

More feedback is welcome also regarding https://github.com/vim/vim/pull/10682#issuecomment-1245797178 and the comments that follow.

**Comment by luukvbaal** [comment]
I updated the first comment with links to the ported vim patches.

File: src/nvim/option_defs.h
Comment (zeertzjq) [inline src/nvim/option_defs.h:737]:
```suggestion
EXTERN int p_spsc;              // 'splitscroll'
```

**Comment by luukvbaal** [comment]
In porting https://github.com/vim/vim/commit/20e58561abc4116f3bfbafaef242d886dd77b303, I took a stab at writing a functional test to replace the screendump test that commit adds. The "Actual" screen state reported by the test does not seem to have `set nosplitscroll` active at all. Running the test sequence manually with `build/bin/nvim -u NONE -S test.vim` achieves the expected result:
```vim
set nosplitscroll
set splitbelow
call setline(1,range(&lines))
function WincmdCb(a, b, c)
	split | wincmd p
endfunction
function TermCb(a, b, c)
	close | split
endfunction
lua << EOF
	vim.api.nvim_set_keymap("n", "j", "", { callback = function()
		vim.cmd("call jobstart([&shell, &shellcmdflag, 'echo'], { 'on_exit': 'WincmdCb' })")
	end})
	vim.api.nvim_set_keymap("n", "t", "", { callback = function()
	vim.api.nvim_set_current_win(
		vim.api.nvim_open_win(vim.api.nvim_create_buf(false, {}), false, { 
			width = 10,
			relative = "cursor",
			height = 4,
			row = 0,
			col = 0,
		}))
		vim.cmd("call termopen(&shell, { 'on_exit': 'TermCb' })")
	end})
EOF
```

Any clue as to what could cause this discrepancy between the screen test and behavior observed when running nvim myself?

**Comment by zeertzjq** [comment]
> Any clue as to what could cause this discrepancy between the screen test and behavior observed when running nvim myself?

I think the difference here is the screen height. If the screen is made a bit taller then the first line of the lower window is no longer 0, although it still doesn't seem correct.

**Comment by luukvbaal** [comment]
Running `build/bin/nvim -u NONE -S test.vim` with the same screen height as the test runs in still yields the expected result.

**Comment by luukvbaal** [comment]
Right, got it working by moving it to it's own function. Wonder what messes with it if left in the `screen_tests()` function.

**Comment by luukvbaal** [comment]
Any particular reason this was moved to the 0.9 milestone btw @clason? Is it too late in the 0.8 cycle? Is anything else needed for this PR? I can keep updating it with patches as vim bug reports/improvements keep coming in but 0.9 seems far away.

**Comment by clason** [comment]
Yes, it's (much) too late; we've been in feature freeze for a week already.

**Comment by luukvbaal** [comment]
> Yes, it's (much) too late; we've been in feature freeze for a week already.

Whoops, totally missed that. Haven't really kept up with neovim development while working on the vim patch😅

Cross posting from https://github.com/vim/vim/pull/10682#issuecomment-1262963386:
> Adding the option to keep the topline the same might actually be simpler to implement.

> >I think it might amount to simply guarding this clause behind the option. If users really want this I could look into it but the option shouldn't get needlessly complicated imo.

So yeah circling back to this after some initial user bug reports/improvements to the feature, I really do think providing this option is as simple as guarding the `winrow != prev_winrow` clause behind an alternate option value. We are then simply (in)validating botline for each window, and correcting the cursor position for curwin.

To reiterate, instead of having the previously proposed options for scroll behavior depending on whether or not the cursor would be valid ({"normal", "noscroll", "neverscroll"}, we dropped this because changing the cursor position was considered not to be that egregious when splitting a window, and as a compromise we add the previous cursor position to the jumplist).

Now the proposal is the have 3 separate scroll behaviors for 'splitscroll' because user(s)(@ychin) might prefer one over the other:
* "normal": default behavior, keep the same relative cursor position in both new and old split.
* "counterscroll": keep the text that was already visible in a window on the same screenlines for each window that is resized/moved by a split command, effectively "counter-scrolling" the topline to keep the text stable. (current 'nosplitscroll' behavior).
* "noscroll": keep the same topline in all windows, effectively only keeping the text stable in the uppermost split.

Not sure "counterscroll" is very telling, can anyone think of a better name for these option values? Do any more users see the need for the third option?

File: src/nvim/testdir/test_window_cmd.vim
Comment (zeertzjq) [inline src/nvim/testdir/test_window_cmd.vim:1507]: Is this because in Vim the splitted window does not have a winbar, while in Nvim it does?

**Review by zeertzjq — APPROVED** [review]
LGTM. I pushed a fixup commit with some small changes.

**Comment by luukvbaal** [comment]
> LGTM. I pushed a fixup commit with some small changes.

Yay! I see I missed some of my own (style) changes from the vim patches. So thanks for the fixup and merge!

**Comment by sibouras** [comment]
this feature is so good pls make a stable 0.81

**Comment by luukvbaal** [comment]
Unfortunately maintainers weren't able to review this before the 0.8 release. I would be willing to make a backport PR but AFAIK 0.x.y releases are only for bug fixes.

**Comment by jacobchrismarsh** [comment]
Any update on if this will get added in the next release?

**Comment by gpanders** [comment]
It will be in the next major release (0.9.0). It is unlikely to be back ported to a point release, as those typically only include bug fixes and, occasionally, very small new features.

**Comment by clason** [comment]
It will be part of the next main release 0.9 but not of maintenance releases like 0.8.1.

**Pull request merged**
